test/unit/logic/delegate.js

```diff
@@ -0,0 +1,522 @@
+
+			delegate.apply(trs, dummyBlock, sender, dummyCb);
+			accountsMock.setAccountAndGet.verify();
+
+			done();
+		});
+	});
+
+	describe('undo', function () {
+
+		var dummyBlock = {
+			id: '9314232245035524467',
+			height: 1
+		};
+
+		it('should call accounts.setAccountAndGet module with correct parameters', function (done) {
+			accountsMock.setAccountAndGet.once().withExactArgs({
+				address: sender.address,
+				u_isDelegate: 1,
+				isDelegate: 0,
+				vote: 0
+			}, dummyCb);
+
+			delegate.undo(trs, dummyBlock, sender, dummyCb);
+			accountsMock.setAccountAndGet.verify();
+
+			done();
+		});
+
+		it('should update username value to null if sender.nameexist is not true', function (done) {
+			delete sender.username;
+			sender.nameexist = 0;
+
+			accountsMock.setAccountAndGet.once().withExactArgs({
+				address: sender.address,
+				u_isDelegate: 1,
+				isDelegate: 0,
+				vote: 0,
+				username: null,
+				u_username: trs.asset.delegate.username
+			}, dummyCb);
+
+			delegate.undo(trs, dummyBlock, sender, dummyCb);
+			accountsMock.setAccountAndGet.verify();
+
+			done();
+		});
+	});
+
+	describe('applyUnconfirmed', function () {
+
+		it('should call accounts.setAccountAndGet module with correct parameters', function (done) {
+			delete sender.username;
+			sender.nameexist = 0;
+
+			accountsMock.setAccountAndGet.once().withExactArgs({
+				address: sender.address,
+				u_isDelegate: 1,
+				isDelegate: 0,
+				username: null,
+				u_username: trs.asset.delegate.username
+			}, dummyCb);
+
+			delegate.applyUnconfirmed(trs, sender, dummyCb);
+			accountsMock.setAccountAndGet.verify();
+
+			done();
+		});
+	});
+
+	describe('undoUnconfirmed', function () {
+
+		it('should call accounts.setAccountAndGet module with correct parameters', function (done) {
+			accountsMock.setAccountAndGet.once().withExactArgs({
+				address: sender.address,
+				u_isDelegate: 0,
+				isDelegate: 0,
+				username: null,
+				u_username: null
+			}, dummyCb);
+
+			delegate.undoUnconfirmed(trs, sender, dummyCb);
+			accountsMock.setAccountAndGet.verify();
+
+			done();
+		});
+	});
+
+	describe('objectNormalize', function () {
+
+		describe('when library.schema.validate fails', function () {
+
+			var schemaDynamicTest = new SchemaDynamicTest({
+				testStyle: SchemaDynamicTest.TEST_STYLE.THROWABLE,
+				customPropertyAssertion: function (input, expectedType, property, err) {
+					expect(err).to.equal('Failed to validate delegate schema: Expected type ' + expectedType + ' but found type ' + input.expectation);
+				}
+			});
+
+			after(function () {
+				describe('schema dynamic tests', function () {
+					schemaDynamicTest.schema.shouldFailAgainst.nonObject.property(delegate.objectNormalize, trs, 'asset.delegate');
+
+					schemaDynamicTest.schema.shouldFailAgainst.nonString.property(delegate.objectNormalize, trs, 'asset.delegate.username');
+				});
+			});
+
+			it('should use the correct format to validate against', function () {
```

Comment:
move the test higher above the first describe block as it should take a place in both - success and failure scenarios